pkg/distributor/distributor.go

```diff
@@ -276,6 +276,8 @@ func (d *Distributor) checkSample(ctx context.Context, userID string, ts client.
 
 // Push implements client.IngesterServer
 func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*client.WriteResponse, error) {
+	defer client.ReuseTS(req.Timeseries)
```

Comment:
All the pushes are very unlikely to all be done when this function finishes. This is due to us returning once a quorum of ingesters have been successfully written to, but we do not cancel the remaining pushes so that our replication factor is still accurate.

Reply:
Good point.  It is highly likely that all the data has been marshalled by this time, but not guaranteed. 
I'd think we do know enough in the tracker to fire the `ReuseTS` when all calls are done, though perhaps an added complication.

Reply:
I suspect this is actually relatively safe - the buf is used to serialise the outgoing request to the ingesters, which is most likely all done by here, even if the request is outstanding.

That being said, its not definitely safe.  You could attach the reuse to the `batchTracker` in `ring`, and have that call a callback when all the RPCs are guaranteed to have finished.